### PR TITLE
(BIDS-1980) switch to balance change for rewards

### PR DESCRIPTION
**Pull request by MauserBitfly**
### <samp>🤖 Generated by Copilot at c4aa858</samp>

This pull request adds support for withdrawals in the validator statistics and earnings feature. It modifies the `validator_stats` table and the queries in `db/statistics.go` to store and calculate the withdrawal data. It also updates the `handlers/common.go` file to display the withdrawal information and the current day's income on the validator page.

### CODE REVIEW

**Review by D13ce — CHANGES_REQUESTED** [review]
Please take a look at my comments.

Also, as discussed, https://github.com/gobitfly/eth2-beaconchain-explorer/blob/c4aa85840bc280f26e9b8efe4703c5d6fa51ffd7/handlers/common.go#L82
can be deleted. This will be done in another PR.

File: db/statistics.go [hunk 1]
```diff
@@ -159,30 +179,55 @@ func WriteValidatorStatisticsForDay(day uint64) error {
...
+			stmt = `
+			INSERT INTO validator_stats (validatorindex, day, cl_rewards_gwei) 
+			(
+				SELECT cur.validatorindex, cur.day,  COALESCE(cur.end_balance, 0) -  COALESCE(cur.start_balance,0) + COALESCE(cur.withdrawals_amount, 0) - COALESCE(cur.deposits_amount,0)  as cl_rewards_gwei
+				FROM validator_stats cur
+				WHERE cur.day = $1 and cur.validatorindex >= $2 and cur.validatorindex <= $3
+			) 
+			ON CONFLICT (validatorindex, day) do
+				update set cl_rewards_gwei = excluded.cl_rewards_gwei;`
+		}
```
[the remark this answers is not in the document]
Comment (MauserBitfly) [reply db/statistics.go]: took your changes

File: handlers/common.go [hunk 2]
```diff
@@ -129,6 +129,47 @@ func GetValidatorEarnings(validators []uint64, currency string) (*types.Validato
...
+	err = db.ReaderDb.Get(&lastBalance, `
+	SELECT 
+        COALESCE(SUM(end_balance), 0) 
+    FROM validator_stats     
+    WHERE day=$2 AND validatorindex = ANY($1)`, validatorsPQArray, lastDay)
+	if err != nil {
+		return nil, nil, err
+	}
+	currentDayIncome := int64(totalBalance - lastBalance - lastDeposits + lastWithdrawals)
+
```
[the remark this answers is not in the document]
Comment (MauserBitfly) [reply handlers/common.go]: Let's think about this in the next PR -> https://github.com/gobitfly/eth2-beaconchain-explorer/pull/2190

File: handlers/common.go [hunk 3]
```diff
@@ -129,6 +129,47 @@ func GetValidatorEarnings(validators []uint64, currency string) (*types.Validato
...
+	var lastBalance uint64
+	err = db.ReaderDb.Get(&lastBalance, `
+	SELECT 
+        COALESCE(SUM(end_balance), 0) 
+    FROM validator_stats     
+    WHERE day=$2 AND validatorindex = ANY($1)`, validatorsPQArray, lastDay)
+	if err != nil {
+		return nil, nil, err
+	}
+	currentDayIncome := int64(totalBalance - lastBalance - lastDeposits + lastWithdrawals)
```
[the remark this answers is not in the document]
Comment (MauserBitfly) [reply handlers/common.go]: renamed

File: handlers/common.go [hunk 4]
```diff
@@ -164,7 +205,7 @@ func GetValidatorEarnings(validators []uint64, currency string) (*types.Validato
 	}
 
 	// retrieve cl income not yet in stats
-	currentDayIncome, currentDayProposerIncome, err := db.GetCurrentDayClIncomeTotal(validators)
+	_, currentDayProposerIncome, err := db.GetCurrentDayClIncomeTotal(validators)
```
[the remark this answers is not in the document]
Comment (MauserBitfly) [reply handlers/common.go]: removed and renamed the function accordingly

**Review by D13ce — APPROVED** [review]
lgtm

**Pull request merged**